### PR TITLE
Fixed remaining case for 'Use local function' refactoring.

### LINKED ISSUE #23149: 'Use local function' breaks code (remaining case)
From @CyrusNajmabadi in https://github.com/dotnet/roslyn/pull/23136#issuecomment-343706220:

> #23136 fixes all of the cases from #22672 except for this one:
>
> ```c#
> using System;
> 
> class Enclosing<T>
> {
>     delegate T MyDelegate(T t);
>     static void Callee(MyDelegate d) => d(default);
> 
>     public class Class<T>
>     {
>         public void Caller()
>         {
>             MyDelegate local = x => x;
>             Callee(local);
>         }
>     }
> }
> ```
>
> We will still offer the feature here, which will produce the broken code:
>
> ```c#
> using System;
> 
> class Enclosing<T>
> {
>     delegate T MyDelegate(T t);
>     static void Callee(MyDelegate d) => d(default);
>
>     public class Class<T>
>     {
>         public void Caller()
>         {
>             T local(T x) => x;
>             Callee(local);
>         }
>     }
> }
> ```
>
> The reason this is broken is that ```local``` is now referencing hte ```T``` type parameter from the containing class ```Class<T>```, whereas before it was using hte ```T``` type parameter from ```Enclosing<T>```.  This breaks because, of course, there's no language facility to reference a specific type parameter with the same name (unlike with types, when you can do things like ```global::```)

### PR DESCRIPTION
**Pull request by zaytsev-victor**
Fixes #23149

<details><summary>Ask Mode template not completed</summary>

### Customer scenario

What does the customer do to get into this situation, and why do we think this
is common enough to address for this release.  (Granted, sometimes this will be
obvious "Open project, VS crashes" but in general, I need to understand how
common a scenario is)

### Bugs this fixes

(either VSO or GitHub links)

### Workarounds, if any

Also, why we think they are insufficient for RC vs. RC2, RC3, or RTW

### Risk

This is generally a measure our how central the affected code is to adjacent
scenarios and thus how likely your fix is to destabilize a broader area of code

### Performance impact

(with a brief justification for that assessment (e.g. "Low perf impact because no extra allocations/no complexity changes" vs. "Low")

### Root cause analysis

How did we miss it?  What tests are we adding to guard against it in the future?

### How was the bug found?

(E.g. customer reported it vs. ad hoc testing)

### Test documentation updated?

If this is a new non-compiler feature or a significant improvement to an existing feature, update https://github.com/dotnet/roslyn/wiki/Manual-Testing once you know which release it is targeting.

</details>

### CODE REVIEW

File: src/Features/CSharp/Portable/UseLocalFunction/CSharpUseLocalFunctionDiagnosticAnalyzer.cs [hunk 1]
```diff
@@ -358,6 +360,39 @@ private void SyntaxNodeAction(SyntaxNodeAnalysisContext syntaxContext, INamedTyp
             return false;
         }
 
+        private static bool CanReplaceDelegateWithLocalFunction(
+            INamedTypeSymbol delegateType, 
+            LocalDeclarationStatementSyntax localDeclaration,
+            SemanticModel semanticModel,
+            CancellationToken cancellationToken)
+        {
+            var delegateTypeParamNames = delegateType.GetAllTypeParameters().Select(p => p.Name).ToImmutableHashSet();
```
Comment (sharwell) [inline src/Features/CSharp/Portable/UseLocalFunction/CSharpUseLocalFunctionDiagnosticAnalyzer.cs]: ❓ Can we fast-path this for the case where the delegate is not contained within a generic type? I believe that would cover the vast majority of delegates, so I'm envisioning this:

```csharp
var delegateContainingType = delegateType.ContainingType;
if (delegateContainingType is null || !delegateContainingType.IsGenericType)
{
    return true;
}
```

⚠️ This is very different from checking if the delegate itself is a generic type, which is quite common.

📝 Due to the age of this pull request, I am only looking for an answer to the question. I plan to update the pull request myself to prepare it for merge.

File: src/EditorFeatures/CSharpTest/UseLocalFunction/UseLocalFunctionTests.cs [hunk 2]
```diff
@@ -1993,6 +1993,76 @@ void Goo()
...
+                Callee(local);
+            }
+        }
+    }
+}");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsUseLocalFunction)]
+        [WorkItem(23149, "https://github.com/dotnet/roslyn/issues/23149")]
+        public async Task TestNotAvaliableIfTypeParameterChanged3()
```
Comment (jcouv) [inline src/EditorFeatures/CSharpTest/UseLocalFunction/UseLocalFunctionTests.cs]: >TestNotAvaliableIfTypeParameterChanged3 [](start = 26, length = 39)

Typo in the test names: Available

File: src/Features/CSharp/Portable/UseLocalFunction/CSharpUseLocalFunctionDiagnosticAnalyzer.cs [hunk 3]
```diff
@@ -358,6 +360,39 @@ private void SyntaxNodeAction(SyntaxNodeAnalysisContext syntaxContext, INamedTyp
...
+            {
+                if (localEnclosingSymbol.Equals(delegateContainingType))
+                {
+                    return true;
+                }
+
+                var typeParams = localEnclosingSymbol.GetTypeParameters();
+                if (typeParams.Any())
+                {
+                    return !typeParams.Any(p => delegateTypeParamNames.Contains(p.Name));
```
Comment (jcouv) [inline src/Features/CSharp/Portable/UseLocalFunction/CSharpUseLocalFunctionDiagnosticAnalyzer.cs]: >return [](start = 20, length = 6)

Feels like this exits too early if there is deep nesting with each layer adding new type parameters.
I would have thought we'd check this layer for problems and if no problem then we'd continue checking containing types.

File: src/Features/CSharp/Portable/UseLocalFunction/CSharpUseLocalFunctionDiagnosticAnalyzer.cs [hunk 4]
```diff
@@ -358,6 +360,39 @@ private void SyntaxNodeAction(SyntaxNodeAnalysisContext syntaxContext, INamedTyp
...
+                }
+
+                localEnclosingSymbol = localEnclosingSymbol.ContainingType;
+            }
+
+            return false;
```
Comment (jcouv) [inline src/Features/CSharp/Portable/UseLocalFunction/CSharpUseLocalFunctionDiagnosticAnalyzer.cs]: I think we should only exit when we reached the type that contains the delegate type. Can this statement be reached? If not, let's add an assertion.

File: src/EditorFeatures/CSharpTest/UseLocalFunction/UseLocalFunctionTests.cs [hunk 5]
```diff
@@ -1993,6 +1993,76 @@ void Goo()
...
+        {
+            await TestMissingAsync(
+@"using System;
+
+class Enclosing<T>
+{
+    delegate T MyDelegate(T t);
+    static void Callee(MyDelegate d) => d(default);
+
+    public class Foo<T>
```
Comment (jcouv) [inline src/EditorFeatures/CSharpTest/UseLocalFunction/UseLocalFunctionTests.cs]: I think we avoid `Foo` in the codebase

File: src/EditorFeatures/CSharpTest/UseLocalFunction/UseLocalFunctionTests.cs [hunk 6]
```diff
@@ -1993,6 +1993,76 @@ void Goo()
...
+    {
+        public class Class
```
Comment (jcouv) [inline src/EditorFeatures/CSharpTest/UseLocalFunction/UseLocalFunctionTests.cs]: Let's add a similar test where the nested type declares some additional type parameters of its own: `class Class<U> ...`

File: src/EditorFeatures/CSharpTest/UseLocalFunction/UseLocalFunctionTests.cs [hunk 7]
```diff
@@ -1993,6 +1993,76 @@ void Goo()
         var buildCancelled = false;
         void onUpdateSolutionCancel(int a) { buildCancelled = true; }
     }
+}");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsUseLocalFunction)]
+        [WorkItem(23149, "https://github.com/dotnet/roslyn/issues/23149")]
+        public async Task TestNotAvaliableIfTypeParameterChanged1()
```
Comment (jcouv) [inline src/EditorFeatures/CSharpTest/UseLocalFunction/UseLocalFunctionTests.cs]: >TestNotAvaliableIfTypeParameterChanged1 [](start = 26, length = 39)

Typo: Available (here and below)

**Comment by jcouv** [comment]
This does not seem ready for another review pass. Previous feedback wasn't addressed yet.

File: src/EditorFeatures/CSharpTest/UseLocalFunction/UseLocalFunctionTests.cs (hunk 2 above)
Comment (sharwell) [reply src/EditorFeatures/CSharpTest/UseLocalFunction/UseLocalFunctionTests.cs]: ➡️ Fixed in 44e67c5796d99ff0dccda995063e0f1092b9cd6f

File: src/Features/CSharp/Portable/UseLocalFunction/CSharpUseLocalFunctionDiagnosticAnalyzer.cs (hunk 3 above)
Comment (sharwell) [reply src/Features/CSharp/Portable/UseLocalFunction/CSharpUseLocalFunctionDiagnosticAnalyzer.cs]: ➡️ Fixed in 44e67c5796d99ff0dccda995063e0f1092b9cd6f with a new test `TestNotAvailableIfTypeParameterChanged4` added

File: src/EditorFeatures/CSharpTest/UseLocalFunction/UseLocalFunctionTests.cs (hunk 6 above)
Comment (sharwell) [reply src/EditorFeatures/CSharpTest/UseLocalFunction/UseLocalFunctionTests.cs]: ➡️ This is the test demonstrating the following issue https://github.com/dotnet/roslyn/pull/23465#discussion_r191032625

File: src/Features/CSharp/Portable/UseLocalFunction/CSharpUseLocalFunctionDiagnosticAnalyzer.cs (hunk 4 above)
Comment (sharwell) [reply src/Features/CSharp/Portable/UseLocalFunction/CSharpUseLocalFunctionDiagnosticAnalyzer.cs]: ➡️ Fixed in 44e67c5796d99ff0dccda995063e0f1092b9cd6f with a new test `TestAvailableIfTypeParameterNotChanged2` added. Also revealed #27950 which is left for future work.

File: src/EditorFeatures/CSharpTest/UseLocalFunction/UseLocalFunctionTests.cs (hunk 5 above)
Comment (sharwell) [reply src/EditorFeatures/CSharpTest/UseLocalFunction/UseLocalFunctionTests.cs]: ➡️ Fixed in 44e67c5796d99ff0dccda995063e0f1092b9cd6f

File: src/EditorFeatures/CSharpTest/UseLocalFunction/UseLocalFunctionTests.cs (hunk 7 above)
Comment (sharwell) [reply src/EditorFeatures/CSharpTest/UseLocalFunction/UseLocalFunctionTests.cs]: ➡️ Fixed in 44e67c5796d99ff0dccda995063e0f1092b9cd6f

**Comment by sharwell** [comment]
@jcouv Feedback addressed and it found three bugs, two of which are now fixed and a third has an issue for it.

**Comment by jcouv** [comment]
Sorry for the delay. I'll finish this review today.

**Comment by jcouv** [comment]
@sharwell If you can resolve the merge conflicts, this is ready to merge. Thanks

**Comment by sharwell** [comment]
retest windows_release_unit64_prtest please

**Pull request merged**